File: newspaper-backend/src/main/java/com/newspaper/service/ArticleService.java
package com.newspaper.service;

import com.newspaper.dto.ArticleDto.*;
import com.newspaper.entity.Article;
import com.newspaper.repository.ArticleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ArticleService {

    private final ArticleRepository articleRepository;

    public PagedArticleResponse getArticles(String keyword, String category, int page, int size, String sortBy) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, sortBy));
        Page<Article> result;

        boolean hasKeyword = keyword != null && !keyword.isBlank();
        boolean hasCategory = category != null && !category.isBlank() && !category.equalsIgnoreCase("All");

        if (hasKeyword && hasCategory) {
            result = articleRepository.searchByKeywordAndCategory(keyword, category, pageable);
        } else if (hasKeyword) {
            result = articleRepository.searchArticles(keyword, pageable);
        } else if (hasCategory) {
            result = articleRepository.findByCategory(category, pageable);
        } else {
            result = articleRepository.findAll(pageable);
        }

        List<ArticleResponse> articles = result.getContent().stream()
                .map(this::toResponse)
                .toList();

        return new PagedArticleResponse(
                articles,
                result.getNumber(),
                result.getTotalPages(),
                result.getTotalElements(),
                result.hasNext(),
                result.hasPrevious()
        );
    }

    public ArticleResponse getArticleById(Long id) {
        Article article = articleRepository.findById(id)
                .orElseThrow(() -> new com.newspaper.exception.ResourceNotFoundException("Article not found with id: " + id));
        return toResponse(article);
    }

    public List<ArticleResponse> getFeaturedArticles() {
        return articleRepository.findByFeaturedTrue().stream()
                .map(this::toResponse)
                .toList();
    }

    public List<String> getCategories() {
        return articleRepository.findDistinctCategories();
    }

    public ArticleResponse createArticle(ArticleRequest request) {
        Article article = Article.builder()
                .title(request.title())
                .content(request.content())
                .excerpt(request.excerpt())
                .category(request.category())
                .author(request.author())
                .imageUrl(request.imageUrl())
                .featured(request.featured())
                .build();
        return toResponse(articleRepository.save(article));
    }

    public ArticleResponse updateArticle(Long id, ArticleRequest request) {
        Article article = articleRepository.findById(id)
                .orElseThrow(() -> new com.newspaper.exception.ResourceNotFoundException("Article not found with id: " + id));
        article.setTitle(request.title());
        article.setContent(request.content());
        article.setExcerpt(request.excerpt());
        article.setCategory(request.category());
        article.setAuthor(request.author());
        article.setImageUrl(request.imageUrl());
        article.setFeatured(request.featured());
        return toResponse(articleRepository.save(article));
    }

    public void deleteArticle(Long id) {
        if (!articleRepository.existsById(id)) {
            throw new com.newspaper.exception.ResourceNotFoundException("Article not found with id: " + id);
        }
        articleRepository.deleteById(id);
    }

    private ArticleResponse toResponse(Article a) {
        return new ArticleResponse(
                a.getId(), a.getTitle(), a.getContent(), a.getExcerpt(),
                a.getCategory(), a.getAuthor(), a.getImageUrl(),
                a.isFeatured(), a.getCreatedAt(), a.getUpdatedAt()
        );
    }
}

File: newspaper-frontend/src/app/services/article.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import {
  Article,
  ArticleRequest,
  PagedArticleResponse
} from '../models/article.model';

@Injectable({ providedIn: 'root' })
export class ArticleService {

  private apiUrl = `${environment.apiUrl}/articles`;

  constructor(private http: HttpClient) {}

  getArticles(
    keyword: string = '',
    category: string = '',
    page: number = 0,
    size: number = 9
  ): Observable<PagedArticleResponse> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    if (keyword) params = params.set('keyword', keyword);
    if (category && category !== 'All') params = params.set('category', category);
    return this.http.get<PagedArticleResponse>(this.apiUrl, { params });
  }

  getArticleById(id: number): Observable<Article> {
    return this.http.get<Article>(`${this.apiUrl}/${id}`);
  }

  getFeaturedArticles(): Observable<Article[]> {
    return this.http.get<Article[]>(`${this.apiUrl}/featured`);
  }

  getCategories(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/categories`);
  }

  createArticle(request: ArticleRequest): Observable<Article> {
    return this.http.post<Article>(this.apiUrl, request);
  }

  updateArticle(id: number, request: ArticleRequest): Observable<Article> {
    return this.http.put<Article>(`${this.apiUrl}/${id}`, request);
  }

  deleteArticle(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
